Index: src/main/java/org/apache/jackrabbit/core/SearchManager.java
===================================================================
--- src/main/java/org/apache/jackrabbit/core/SearchManager.java	(revision 576874)
+++ src/main/java/org/apache/jackrabbit/core/SearchManager.java	(working copy)
@@ -296,6 +296,10 @@
         return query;
     }
 
+    public QueryHandler getQueryHandler(){
+        return handler;
+    }
+    
     /**
      * Creates a prepared query object that can be executed on the workspace.
      *
Index: src/main/java/org/apache/jackrabbit/core/RepositoryImpl.java
===================================================================
--- src/main/java/org/apache/jackrabbit/core/RepositoryImpl.java	(revision 576874)
+++ src/main/java/org/apache/jackrabbit/core/RepositoryImpl.java	(working copy)
@@ -799,7 +799,7 @@
      * @throws RepositoryException      if an error occurs while opening the
      *                                  search index.
      */
-    SearchManager getSearchManager(String workspaceName)
+    public SearchManager getSearchManager(String workspaceName)
             throws NoSuchWorkspaceException, RepositoryException {
         // check sanity of this instance
         sanityCheck();
