Index: wicket/src/main/java/org/apache/wicket/ajax/wicket-ajax.js
===================================================================
--- wicket/src/main/java/org/apache/wicket/ajax/wicket-ajax.js	(revision 1071587)
+++ wicket/src/main/java/org/apache/wicket/ajax/wicket-ajax.js	(working copy)
@@ -126,6 +126,7 @@
 	}
 },
 
+
 /**
  * Functions executer takes array of functions and executes them. Each function gets
  * the notify object, which needs to be called for the next function to be executed.
@@ -172,6 +173,12 @@
 	}
 }
 
+/*
+  WICKET-3473 Helper to proceed to next step, execution of notify needs to be delayed until the head contributions are loaded
+*/
+Wicket.functionExecuterSeq = 0;
+Wicket.functionExecuterCallbacks = {};
+
 Wicket.replaceOuterHtmlIE = function(element, text) {						
 
 	// replaces all <iframe references with <__WICKET_JS_REMOVE_X9F4A__iframe text
@@ -1624,12 +1631,13 @@
 				// load the external javascript using Wicket.Ajax.Request
 				
 				// callback when script is loaded
-				var onLoad = function(content) {					
-					Wicket.Head.addJavascript(content, null, src);
-					Wicket.Ajax.invokePostCallHandlers();
-
-					// continue to next step
-					notify();
+                var callBackIdentifier = 'script' + (Wicket.functionExecuterSeq++);
+                var onLoad = function(content) {
+                    Wicket.functionExecuterCallbacks[callBackIdentifier] = function() {
+                        Wicket.Ajax.invokePostCallHandlers();
+                        notify();
+                    }
+					Wicket.Head.addJavascript(content+"; Wicket.functionExecuterCallbacks['"+callBackIdentifier+"'](); delete Wicket.functionExecuterCallbacks['"+callBackIdentifier+"']; ", null, src);
 				}
 				// we need to schedule the request as timeout
 				// calling xml http request from another request call stack doesn't work
@@ -2421,4 +2429,4 @@
 	iframe.style.display="none";
 	iframe.style.visibility="hidden";
 	return iframe;
-}
\ No newline at end of file
+}
