Index: src/main/webapp/xinha/plugins/TableOperations/lang/nl.js
===================================================================
--- src/main/webapp/xinha/plugins/TableOperations/lang/nl.js	(revision 1249)
+++ src/main/webapp/xinha/plugins/TableOperations/lang/nl.js	(working copy)
@@ -9,7 +9,7 @@
   "Border": "Rand",
   "Borders": "Randen",
   "Bottom": "Onder",
-  "Style [CSS]": "CSS Style",
+  "Style [CSS]": "CSS Stijl",
   "Caption": "Opmerking",
   "Cell Properties": "Celeigenschappen",
   "Center": "Centreren",
@@ -19,7 +19,7 @@
   "Description": "Omschrijving",
   "FG Color": "Voorgrond",
   "Float": "Zwevend",
-  "Frames": "Frames",
+  "Frames": "Zijlijnen",
   "Height": "Hoogte",
   "How many columns would you like to merge?": "Hoeveel kolommen wilt u samenvoegen?",
   "How many rows would you like to merge?": "Hoeveel rijen wilt u samenvoegen?",
@@ -41,7 +41,7 @@
   "Rules will appear between rows only": "Regels verschijnen enkel tussen de rijen",
   "Rules": "Regels",
   "Spacing and padding": "Celmarge en afstand tussen cellen",
-  "Spacing": "marge",
+  "Spacing": "Marge",
   "Summary": "Overzicht",
   "Delete cell": "Cel verwijderen",
   "Insert cell after": "Voeg cel toe achter",
@@ -69,7 +69,7 @@
   "The top side only": "Enkel aan de bovenkant",
   "Top": "Boven",
   "Unset color": "Wis kleur",
-  "Vertical align": "Vertikale uitlijning",
+  "Vertical align": "Verticale uitlijning",
   "Width": "Breedte",
   "Xinha cowardly refuses to delete the last cell in row.": "Xinha kan de laatste cel in deze tabel niet verwijderen.",
   "Xinha cowardly refuses to delete the last column in table.": "Xinha kan de laatste kolom in deze tabel niet verwijderen.",
Index: src/main/webapp/xinha/plugins/TableOperations/popups/dialogTable.html
===================================================================
--- src/main/webapp/xinha/plugins/TableOperations/popups/dialogTable.html	(revision 1249)
+++ src/main/webapp/xinha/plugins/TableOperations/popups/dialogTable.html	(working copy)
@@ -1,5 +1,5 @@
  <h1 id="[h1]"><l10n>Table Properties</l10n></h1>
-  <table style="width: 100%;">
+  <table style="width: 100%;text-align:left;">
       <tbody>
           <tr>
               <td>
@@ -7,7 +7,7 @@
                       <legend>
                           <l10n>Description</l10n>
                       </legend>
-                      <table style="width: 100%;">
+                      <table style="width: 98%;">
                           <tbody>
                               <tr>
                                   <td class="label">
@@ -40,7 +40,7 @@
                       <legend>
                           <l10n>Spacing and padding</l10n>
                       </legend>
-                      <table style="width: 100%;">
+                      <table style="width: 98%;">
                           <tbody>
                               <tr>
                                   <td class="label">
@@ -61,7 +61,7 @@
                       <legend>
                           <l10n>Frame and borders</l10n>
                       </legend>
-                      <table width="100%">
+                      <table width="98%">
                           <tbody>
                               <tr>
                                   <td class="label">
@@ -90,7 +90,7 @@
                               </tr>
                               <tr>
                                   <td class="label">
-                                      <l10n>Rules:
+                                      <l10n>Rules</l10n>:
                                   </td>
                                   <td>
                                       <select name="[rules]">
Index: src/main/webapp/xinha/lang/nl.js
===================================================================
--- src/main/webapp/xinha/lang/nl.js	(revision 1249)
+++ src/main/webapp/xinha/lang/nl.js	(working copy)
@@ -32,7 +32,32 @@
   "Horizontal Rule": "Horizontale lijn",
   "Insert Web Link": "Hyperlink invoegen/aanpassen",
   "Insert/Modify Image": "Afbeelding invoegen/aanpassen",
+  // Insert Table
   "Insert Table": "Tabel invoegen",
+  "Caption": "Opmerking",
+  "Rows:": "Rijen:",
+  "Number of rows": "Aantal rijen",
+  "Cols:": "Kolommen:",
+  "Number of columns": "Aantal kolommen",
+  "Width:": "Breedte:",
+  "Width of the table": "Tabelbreedte",
+  "Percent": "Procent",
+  "Pixels": "Pixels",
+  "Em": "Em",
+  "Layout": "Opmaak",
+  "Spacing": "Marges",
+  "Alignment:": "Uitlijning:",
+  "Border": "Rand",
+  "Collapse borders:": "Geen randen:",
+  "Width unit": "Breedte eenheid",
+  "Fixed width columns": "Vaste breedte voor kolommen",
+  "Positioning of this table": "Positionering van de tabel",
+  "Cell spacing:": "tussen cellen:",
+  "Space between adjacent cells": "Ruimte tussen aangrenzende cellen",
+  "Cell padding:": "binnen cellen:",
+  "Space between content and border in cell": "Ruimte tusse de inhoud en de rand van een cel",
+  "You must enter a number of rows": "Aantal rijen is verplicht",
+  "You must enter a number of columns": "Aantal kolommen is verplicht",
   "Toggle HTML Source": "HTML broncode",
   "Enlarge Editor": "Vergroot Editor",
   "About this editor": "Over deze editor",
Index: src/main/webapp/xinha/modules/InlineStyler/InlineStyler.js
===================================================================
--- src/main/webapp/xinha/modules/InlineStyler/InlineStyler.js	(revision 1249)
+++ src/main/webapp/xinha/modules/InlineStyler/InlineStyler.js	(working copy)
@@ -342,7 +342,7 @@
   input.value = Xinha._colorToRgb( el.style.backgroundColor );
   input.type = "hidden";
   this.inputs.styles['backgroundColor'] = input;
-  input.style.marginRight = "0.5em";
+  input.style.marginRight = "3.5em";
   td.appendChild(input);
   new Xinha.colorPicker.InputBinding(input)
   
Index: src/main/webapp/xinha/modules/InsertTable/dialog.html
===================================================================
--- src/main/webapp/xinha/modules/InsertTable/dialog.html	(revision 1249)
+++ src/main/webapp/xinha/modules/InsertTable/dialog.html	(working copy)
@@ -16,7 +16,7 @@
   <tr>
     <td style="width: 4em; text-align: right"><l10n>Rows:</l10n></td>
     <td><input type="text" name="[rows]" id="[rows]" size="5" title="_(Number of rows)"  /></td>
-    <td style="width: 4em; text-align: right">Width:</td>
+    <td style="width: 4em; text-align: right"><l10n>Width:</l10n></td>
     <td><input type="text" name="[width]" id="[width]" size="5" title="_(Width of the table)" /></td>
     <td><select size="1" name="[unit]" id="[unit]" title="Width unit">
       <option value="%"               ><l10n>Percent</l10n></option>
@@ -25,7 +25,7 @@
     </select></td>
   </tr>
   <tr>
-    <td style="width: 4em; text-align: right">Cols:</td>
+    <td style="width: 4em; text-align: right"><l10n>Cols:</l10n></td>
     <td><input type="text" name="[cols]" id="[cols]" size="5" title="_(Number of columns)"  /></td>
     <td style="text-align: right"><input type="checkbox" name="[fixed]" id="[fixed]" value="on" /></td>
     <td colspan="2"><label
@@ -37,7 +37,7 @@
 <br />
 
 <fieldset id="[layout_fieldset]" style="float: left; margin-left: 10px;">
-<legend><l10n>Layou</l10n>t</legend>
+<legend><l10n>Layout</l10n></legend>
 
 <div class="space"></div>
 
@@ -57,6 +57,7 @@
 </select>

 <br />
+  <div class="space"></div>

 <div class="fl"><l10n>Border</l10n></div>
 <input type="text" name="[border]" id="[border]" size="2" title="_(Leave empty for no border)" /> <l10n>Pixels</l10n>
@@ -71,7 +72,9 @@
   <option value="outset"  >outset</option>
 </select>
 <input type="text" size="7" id="[border_color]" name="[border_color]"/>
-<br /><l10n>Collapse borders:</l10n>
+<br />
+  <div class="space"></div>
+  <l10n>Collapse borders:</l10n>
 <input type="checkbox" name="[border_collapse]" id="[border_collapse]" value="on" />

 <div class="space"></div>
@@ -87,6 +90,7 @@
 <input type="text" name="[spacing]" id="[spacing]" size="5" title="_(Space between adjacent cells)" />

 <br />
+<div class="space"></div>

 <div class="fr"><l10n>Cell padding:</l10n></div>
 <input type="text" name="[padding]" id="[padding]" size="5" title="_(Space between content and border in cell)" />
Index: src/main/webapp/xinha/modules/InsertTable/insert_table.js
===================================================================
--- src/main/webapp/xinha/modules/InsertTable/insert_table.js	(revision 1249)
+++ src/main/webapp/xinha/modules/InsertTable/insert_table.js	(working copy)
@@ -73,7 +73,7 @@
 	var self = this;
 	var editor = this.editor;

-	var dialog = this.dialog = new Xinha.Dialog(editor, InsertTable.html, 'Xinha',{width:400})
+	var dialog = this.dialog = new Xinha.Dialog(editor, InsertTable.html, 'Xinha',{width:520})
 	// Connect the OK and Cancel buttons
 	dialog.getElementById('ok').onclick = function() {self.apply();}
 	dialog.getElementById('cancel').onclick = function() { self.dialog.hide()};
