Index: src/main/webapp/xinha/plugins/InsertAnchor/InsertAnchor.js
===================================================================
--- src/main/webapp/xinha/plugins/InsertAnchor/InsertAnchor.js	(revision 1249)
+++ src/main/webapp/xinha/plugins/InsertAnchor/InsertAnchor.js	(working copy)
@@ -85,7 +85,7 @@
 	var editor = this.editor;
 	this.selectedHTML = editor.getSelectedHTML();
 	var sel  = editor.getSelection();
-  var range  = editor.createRange(sel);
+  this.range  = editor.createRange(sel);
   this.a = editor.activeElement(sel);
   
   if(!(this.a != null && this.a.tagName.toLowerCase() == 'a'))
@@ -135,11 +135,10 @@
           a.title = anchor;
           a.className = "anchor";
           a.innerHTML = self.placeholderImg;
-			var html = editor.getSelectedHTML();
-		  if (html) a.innerHTML += html;
+		  if (this.selectedHTML) a.innerHTML += this.selectedHTML;
 			if (Xinha.is_ie) 
 			{
-            range.pasteHTML(a.outerHTML);
+            this.range.pasteHTML(a.outerHTML);
           	}
 			else 
 			{
